corehq/motech/repeaters/tasks.py

```diff
@@ -72,65 +64,103 @@ def delete_old_request_logs():
     queue=settings.CELERY_PERIODIC_QUEUE,
 )
 def check_repeaters():
-    # this creates a task for all partitions
-    # the Nth child task determines if a lock is available for the Nth partition
-    for current_partition in range(CHECK_REPEATERS_PARTITION_COUNT):
-        check_repeaters_in_partition.delay(current_partition)
-
-
-@task(queue=settings.CELERY_PERIODIC_QUEUE)
-def check_repeaters_in_partition(partition):
-    """
-    The CHECK_REPEATERS_PARTITION_COUNT constant dictates the total number of partitions
-    :param partition: index of partition to check
-    """
     start = datetime.utcnow()
     twentythree_hours_sec = 23 * 60 * 60
     twentythree_hours_later = start + timedelta(hours=23)
 
-    # Long timeout to allow all waiting repeat records to be iterated
-    lock_key = f"{CHECK_REPEATERS_KEY}_{partition}_in_{CHECK_REPEATERS_PARTITION_COUNT}"
+    # Long timeout to allow all waiting repeaters to be iterated
+    # TODO: Check how long it takes to iterate all repeaters on prod
     check_repeater_lock = get_redis_lock(
-        lock_key,
+        CHECK_REPEATERS_KEY,
         timeout=twentythree_hours_sec,
-        name=lock_key,
+        name=CHECK_REPEATERS_KEY,
     )
     if not check_repeater_lock.acquire(blocking=False):
-        metrics_counter("commcare.repeaters.check.locked_out", tags={'partition': partition})
+        metrics_counter("commcare.repeaters.check.locked_out")
         return
 
     try:
         with metrics_histogram_timer(
             "commcare.repeaters.check.processing",
             timing_buckets=_check_repeaters_buckets,
         ):
-            for record in RepeatRecord.objects.iter_partition(
-                    start, partition, CHECK_REPEATERS_PARTITION_COUNT):
-                if not _soft_assert(
-                    datetime.utcnow() < twentythree_hours_later,
-                    "I've been iterating repeat records for 23 hours. I quit!"
-                ):
-                    break
-
+            for repeater in iter_ready_repeaters():
                 metrics_counter("commcare.repeaters.check.attempt_forward")
-                record.attempt_forward_now(is_retry=True)
-            else:
-                iterating_time = datetime.utcnow() - start
-                _soft_assert(
-                    iterating_time < timedelta(hours=6),
-                    f"It took {iterating_time} to iterate repeat records."
-                )
+                process_repeater.delay(repeater.domain, repeater.repeater_id)
+
+                if datetime.utcnow() > twentythree_hours_later:
+                    # Break after process_repeater() to avoid a stale lock
+                    break
     finally:
         check_repeater_lock.release()
 
 
+def iter_ready_repeaters():
+    """
+    Cycles through repeaters (repeatedly ;) ) until there are no more
+    repeat records ready to be sent.
+    """
+    while True:
+        yielded = False
+        for repeater in Repeater.objects.all_ready():
```

Comment:
Maybe a naive suggestion, but I looked at this again with Evan and was wondering why you couldn't "lock" a repeater by setting a property on the Repeater itself, something like `is_processing` and only return repeaters where that is False. I imagine the two cons are:
- this isn't actually a lock, so if somehow two `process_repeater` calls were made for the same repeater, it would still process simultaneously
- adds another database update on the repeaters table

Reply:
I was thinking the same thing. And I think we can avoid the first con by setting the property in the iterator.

Reply:
Locking that sucks less implemented here: 2bb7c595054